### PR TITLE
Feature/duplication

### PR DESCRIPTION
**Pull request by cini9**

### CODE REVIEW

File: api/app/models/instance.rb [hunk 1]
```diff
@@ -29,13 +29,13 @@ class Instance < ApplicationRecord
   validates :instanceable_type, inclusion: { in: %w(Algorithm DecisionTree Node) }
   validates_uniqueness_of :node_id, scope: [:instanceable_id, :instanceable_type, :diagnosis_id]
 
-  validates :duration_translations, translated_fields_presence: { project: lambda { |record|
-    record.node.project_id
-  } }, unless: Proc.new { !node.is_a?(HealthCares::Drug) || is_pre_referral  }
+  # validates :duration_translations, translated_fields_presence: { project: lambda { |record|
```
Comment (qtn-g) [inline api/app/models/instance.rb]: ?

File: api/app/services/duplicate_algorithm_service.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        run_function(I18n.t('algorithms.duplication.duplicating_diagram')) { duplicate_diagram }
+        run_function(I18n.t('algorithms.duplication.duplicating_medal_data_variables')) { duplicate_medal_data_variables }
+        run_function(I18n.t('algorithms.duplication.duplicating_decision_trees')) { duplicate_decision_trees }
+        run_function(I18n.t('algorithms.duplication.duplicating_exclusions')) { duplicate_exclusions }
+        run_function(I18n.t('algorithms.duplication.adjust_instances')) { adjust_diagnoses_instances }
+        run_function(I18n.t('algorithms.duplication.end_duplication'), 'finished')
+        remove_history_file
+        # Algorithm.validate_duplicate(self, new_algorithm)
+      rescue => e
+        puts "############################################################"
```
Comment (qtn-g) [inline api/app/services/duplicate_algorithm_service.rb]: remove

File: api/app/services/duplicate_algorithm_service.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+        run_function(I18n.t('algorithms.duplication.start_duplication'), 'starting')
+        run_function(I18n.t('algorithms.duplication.duplicating_diagram')) { duplicate_diagram }
+        run_function(I18n.t('algorithms.duplication.duplicating_medal_data_variables')) { duplicate_medal_data_variables }
+        run_function(I18n.t('algorithms.duplication.duplicating_decision_trees')) { duplicate_decision_trees }
+        run_function(I18n.t('algorithms.duplication.duplicating_exclusions')) { duplicate_exclusions }
+        run_function(I18n.t('algorithms.duplication.adjust_instances')) { adjust_diagnoses_instances }
+        run_function(I18n.t('algorithms.duplication.end_duplication'), 'finished')
+        remove_history_file
+        # Algorithm.validate_duplicate(self, new_algorithm)
```
Comment (qtn-g) [inline api/app/services/duplicate_algorithm_service.rb]: Why is it commented ?

File: api/app/services/duplicate_algorithm_service.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        @previous_message = ''
+
+        DecisionTree.skip_callback(:create, :after, :generate_reference, raise: false)
+        Node.skip_callback(:create, :after, :generate_reference, raise: false)
+        Diagnosis.skip_callback(:create, :after, :instantiate_in_diagram, raise: false)
+
+        # Recreate version
+        @new_algorithm = Algorithm.create!(
+          @algorithm.attributes.except('id', 'name', 'status', 'job_id', 'created_at', 'updated_at')
+                    .merge({'name': "Copy of #{@algorithm.name}", 'status': 'draft'}))
```
Comment (qtn-g) [inline api/app/services/duplicate_algorithm_service.rb]: Translate it with the project default language

File: front/components/duplicate.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * The internal imports
+ */
+import Card from '@/components/card'
+import { useGetAlgorithmQuery } from '@/lib/api/modules/enhanced/algorithm.enhanced'
+import { customFormatDuration } from '@/lib/utils/date'
+import { useToast } from '@/lib/hooks/useToast'
+import { useWebSocket } from '@/lib/hooks/useWebSocket'
+import { DuplicateComponent } from '@/types'
```
Comment (qtn-g) [inline front/components/duplicate.tsx]: type

File: front/components/duplicate.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    elementId ? { id: elementId } : skipToken
+  )
+
+  /**
+   * Queue toast if error during destruction
+   */
+  useEffect(() => {
+    if (isAlgorithmError) {
+      newToast({
+        message: t('notifications.archiveError', { ns: 'common' }),
```
Comment (qtn-g) [inline front/components/duplicate.tsx]: update messages

File: front/components/duplicate.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    elementId,
+    error: webSocketError,
+  } = useWebSocket()
+
+  const { data: algorithm, isError: isAlgorithmError } = useGetAlgorithmQuery(
+    elementId ? { id: elementId } : skipToken
+  )
+
+  /**
+   * Queue toast if error during destruction
```
Comment (qtn-g) [inline front/components/duplicate.tsx]: remove or change it

File: front/components/duplicate.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                  : t('noDuplication')}
+              </Text>
+            </HStack>
+            <AccordionIcon />
+          </AccordionButton>
+          <AccordionPanel p={algorithm ? 4 : 0}>
+            <VStack alignItems='flex-start' w='full'>
+              {messages &&
+                messages.map(message => (
+                  <HStack justifyContent='space-between' w='full'>
```
Comment (qtn-g) [inline front/components/duplicate.tsx]: 🤦

File: front/lib/hooks/useChannel.ts [hunk 9]
```diff
@@ -47,10 +47,23 @@ export function useChannel(actionCable: Cable) {
     }
   }
 
+  const getCurrentJobInfo = () => {
+    if (subscribed && !connected)
+      throw new Error('useActionCable: not connected')
+    if (!subscribed) throw new Error('useActionCable: not subscribed')
+    try {
+      console.log('Requesting current job info')
```
Comment (qtn-g) [inline front/lib/hooks/useChannel.ts]: remove

File: front/pages/projects/[projectId]/algorithms/index.tsx [hunk 10]
```diff
@@ -106,13 +123,13 @@ export default function Algorithms() {
    * Queue toast if error during destruction
    */
   useEffect(() => {
-    if (isDestroyError) {
+    if (isDestroyError || isDuplicateError) {
       newToast({
         message: t('notifications.archiveError', { ns: 'common' }),
```
Comment (qtn-g) [inline front/pages/projects/[projectId]/algorithms/index.tsx]: wrong translations

File: front/pages/projects/[projectId]/algorithms/index.tsx [hunk 11]
```diff
@@ -145,40 +162,45 @@ export default function Algorithms() {
               itemId={row.id}
               onEdit={() => onEdit(row.id)}
               onArchive={
-                row.status !== 'archived' && isAdminOrClinician
-                  ? () => onArchive(row.id)
-                  : undefined
+                row.status !== 'archived' ? () => onArchive(row.id) : undefined
```
Comment (qtn-g) [inline front/pages/projects/[projectId]/algorithms/index.tsx]: use algoritm status enum

File: front/types/component.d.ts [hunk 12]
```diff
@@ -158,4 +160,15 @@ export type FormLabelComponent = FC<
...
 
+export type DuplicateComponent = FC<{
+  error:
+    | ClientError
+    | {
+        message: string
+      }
+    | SerializedError
+    | undefined
+  setIsDuplicating: React.Dispatch<React.SetStateAction<boolean>>
```
Comment (qtn-g) [inline front/types/component.d.ts]: React is missing in external imports

**Review by qtn-g — CHANGES_REQUESTED** [review]

File: api/app/models/instance.rb (hunk 1 above)
Comment (cini9) [reply api/app/models/instance.rb:32]: @ManuBarchich ?

File: api/app/graphql/mutations/algorithms/duplicate_algorithm.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        project_id = algorithm.project.id
+        return true if context[:current_api_v2_user].project_clinician?(project_id)
+
+        raise GraphQL::ExecutionError, I18n.t('graphql.errors.wrong_access', class_name: 'Project')
+      end
+
+      # Resolve
+      def resolve(id:)
+        begin
+          DuplicateAlgorithmJob.perform_now(id)
```
Comment (frescoal) [inline api/app/graphql/mutations/algorithms/duplicate_algorithm.rb:22]: Perform_later nan ?

File: front/components/duplicate.tsx (hunk 8 above)
Comment (cini9) [reply front/components/duplicate.tsx]: 🤫

File: front/components/publish.tsx [hunk 14]
```diff
@@ -80,7 +80,7 @@ const Publish = () => {
     }
   }, [isError])
 
-  const isSingleValue = (value: any): value is SingleValue<Option> =>
+  const isSingleValue = (value: unknown): value is SingleValue<Option> =>
```
Comment (frescoal) [inline front/components/publish.tsx:83]: Is that normal ?

**Review by frescoal — CHANGES_REQUESTED** [review]
Les tests passent pas

File: front/components/publish.tsx (hunk 14 above)
Comment (cini9) [reply front/components/publish.tsx:83]: It is, I figured unknown was more appropriate than any

**Pull request merged**